File: Practice/20200302/ProjectEuler-10/10.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <numeric>

void isPrime(int n, std::vector<int> &primes)
{
    int top = std::floor(std::sqrt(n));
    for (auto i = primes.begin(); i != primes.end(); ++i)
    {
        if (*i > top)
        {
            primes.push_back(n);
            return;
        }
        if (n % *i == 0)
            return;
    }
}

int main()
{
    size_t top;
    std::cout << "Upper limit: ";
    std::cin >> top;
    std::vector<int> primes{2, 3};
    for (size_t i = 4; i <= top; ++i)
    {
        isPrime(i, primes);
    }
    std::cout << primes.size() << std::endl;
    for (auto i = primes.begin(); i != primes.end(); ++i)
        std::cout << *i << ' ';
    std::cout << std::endl;
    unsigned long long int sum = 0;
    for (auto i = primes.begin(); i != primes.end(); ++i)
        sum += *i;
    // sum = std::accumulate(primes.begin(), primes.end(), 0);
    std::cout << sum << std::endl;
}

File: Practice/20200302/ProjectEuler-10/10.py
#!/usr/bin/env python3
import math


def is_prime(n, known_primes):
    top = math.floor(math.sqrt(n))
    for i in (x for x in known_primes if x <= top):
        if not n % i:
            return
    known_primes.append(n)


def main():
    end = int(input('Upper limit: '))
    primes = [2, 3]
    try:
        for n in range(4, end):
            is_prime(n, primes)
    except KeyboardInterrupt:
        print(n)
    # print(primes)
    print(len(primes))
    print(sum(primes))


if __name__ == "__main__":
    res = main()
